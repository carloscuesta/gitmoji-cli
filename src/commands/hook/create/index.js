// @flow
import fs from 'fs'
import execa from 'execa'
import ora from 'ora'

import HOOK from '../hook'

const createHook = async () => {
  const spinner = ora('Creating the gitmoji commit hook').start()

  try {
    const { stdout } = await execa('git', ['rev-parse', '--absolute-git-dir'])

    fs.writeFile(
      stdout + HOOK.PATH,
      HOOK.CONTENTS,
      { mode: HOOK.PERMISSIONS },
      (error) => {
        if (error) return spinner.fail(error)
        spinner.succeed('Gitmoji commit hook created successfully')
      }
    )
  } catch (error) {
    spinner.fail(`Error: ${error}`)
  }
}

export default createHook
