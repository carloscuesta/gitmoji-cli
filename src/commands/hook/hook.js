// @flow
const HOOK: Object = {
  PERMISSIONS: 0o775,
  PATH: '/hooks/prepare-commit-msg',
  CONTENTS: `#!/bin/sh
# gitmoji as a commit hook
commit_message_empty=true
while read -r line
do
  case "$line" in
  "# ------------------------ >8 ------------------------"*)
    break
    ;;
  ""|"#"*)
    # skip
    ;;
  *)
    commit_message_empty=false
    break
    ;;
  esac
done < "$1"
if test -t 1 && $commit_message_empty; then
  # it has been invoked from a tty and no commit message is already set
  exec < /dev/tty
  gitmoji --hook "$1"
fi`
}

export default HOOK
