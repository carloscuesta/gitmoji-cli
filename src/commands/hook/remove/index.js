// @flow
import fs from 'fs'
import execa from 'execa'
import ora from 'ora'

import HOOK from '../hook'

const removeHook = async () => {
  const spinner = ora('Creating the gitmoji commit hook').start()

  try {
    const { stdout } = await execa('git', ['rev-parse', '--absolute-git-dir'])

    fs.unlink(stdout + HOOK.PATH, (error) => {
      if (error)
        return spinner.fail('Error: Gitmoji commit hook is not created')
      spinner.succeed('Gitmoji commit hook removed successfully')
    })
  } catch (error) {
    spinner.fail(`Error: ${error}`)
  }
}

export default removeHook
