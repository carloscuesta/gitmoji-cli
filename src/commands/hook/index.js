import create from './create'
import remove from './remove'

export default {
  create,
  remove
}
