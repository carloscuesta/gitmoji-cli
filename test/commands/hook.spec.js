import execa from 'execa'
import fs from 'fs'

import hook from '../../src/commands/hook'
import hookConfig from '../../src/commands/hook/hook'
import * as stubs from './stubs'

describe('hook command', () => {
  it('should match hook module export', () => {
    expect(hook).toMatchSnapshot()
  })

  it('should match hook configuration', () => {
    expect(hookConfig).toMatchSnapshot()
  })

  describe('create hook', () => {
    beforeAll(() => {
      execa.mockReturnValue({ stdout: stubs.gitAbsoluteDir })
      hook.create()
    })

    it('should obtain the absolute git dir with execa', () => {
      expect(execa).toHaveBeenCalledWith('git', [
        'rev-parse',
        '--absolute-git-dir'
      ])
    })

    it('should create the hook file', () => {
      expect(fs.writeFile).toHaveBeenCalledWith(
        stubs.gitAbsoluteDir + hookConfig.PATH,
        hookConfig.CONTENTS,
        { mode: hookConfig.PERMISSIONS },
        expect.any(Function)
      )
    })
  })

  describe('remove hook', () => {
    beforeAll(() => {
      execa.mockReturnValue({ stdout: stubs.gitAbsoluteDir })
      hook.remove()
    })

    it('should obtain the absolute git dir with execa', () => {
      expect(execa).toHaveBeenCalledWith('git', [
        'rev-parse',
        '--absolute-git-dir'
      ])
    })

    it('should remove the hook file', () => {
      expect(fs.unlink).toHaveBeenCalledWith(
        stubs.gitAbsoluteDir + hookConfig.PATH,
        expect.any(Function)
      )
    })
  })
})
