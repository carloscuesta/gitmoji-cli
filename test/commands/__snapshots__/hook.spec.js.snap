// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hook command should match hook configuration 1`] = `
Object {
  "CONTENTS": "#!/bin/sh
# gitmoji as a commit hook
commit_message_empty=true
while read -r line
do
  case \\"$line\\" in
  \\"# ------------------------ >8 ------------------------\\"*)
    break
    ;;
  \\"\\"|\\"#\\"*)
    # skip
    ;;
  *)
    commit_message_empty=false
    break
    ;;
  esac
done < \\"$1\\"
if test -t 1 && $commit_message_empty; then
  # it has been invoked from a tty and no commit message is already set
  exec < /dev/tty
  gitmoji --hook \\"$1\\"
fi",
  "PATH": "/hooks/prepare-commit-msg",
  "PERMISSIONS": 509,
}
`;

exports[`hook command should match hook module export 1`] = `
Object {
  "create": [Function],
  "remove": [Function],
}
`;
